document-parsers/langchain4j-document-parser-apache-tika/src/main/java/dev/langchain4j/data/document/parser/apache/tika/ApacheTikaDocumentParser.java

```diff
@@ -0,0 +1,73 @@
+import org.apache.tika.metadata.Metadata;
+import org.apache.tika.parser.AutoDetectParser;
+import org.apache.tika.parser.ParseContext;
+import org.apache.tika.parser.Parser;
+import org.apache.tika.sax.BodyContentHandler;
+import org.xml.sax.ContentHandler;
+
+import java.io.InputStream;
+
+import static dev.langchain4j.internal.Utils.getOrDefault;
+
+/**
+ * Parses files into {@link Document}s using Apache Tika library, automatically detecting the file format.
+ * This parser supports various file formats, including PDF, DOC, PPT, XLS.
+ * For detailed information on supported formats,
+ * please refer to the <a href="https://tika.apache.org/2.9.1/formats.html">Apache Tika documentation</a>.
+ */
+public class ApacheTikaDocumentParser implements DocumentParser {
+
+    private static final int UNLIMITED = -1;
+
+    private final Parser parser;
+    private final ContentHandler contentHandler;
+    private final Metadata metadata;
+    private final ParseContext parseContext;
+
+    /**
+     * Creates an instance of {@code ApacheTikaDocumentParser} with the default Tika components.
+     * It uses {@link AutoDetectParser}, {@link BodyContentHandler} without write limit,
+     * empty {@link Metadata} and empty {@link ParseContext}.
+     */
+    public ApacheTikaDocumentParser() {
+        this(null, null, null, null);
+    }
+
+    /**
+     * Creates an instance of {@code ApacheTikaDocumentParser} with the provided Tika components.
+     * If some of the components is not provided ({@code null}, the default will be used.
+     * It uses {@link AutoDetectParser}, {@link BodyContentHandler} without write limit,
+     * empty {@link Metadata} and empty {@link ParseContext}.
+     *
+     * @param parser         Tika parser to use. Default: {@link AutoDetectParser}
+     * @param contentHandler Tika content handler. Default: {@link BodyContentHandler} without write limit
+     * @param metadata       Tika metadata. Default: empty {@link Metadata}
+     * @param parseContext   Tika parse context. Default: empty {@link ParseContext}
+     */
+    public ApacheTikaDocumentParser(Parser parser,
+                                    ContentHandler contentHandler,
+                                    Metadata metadata,
+                                    ParseContext parseContext) {
+        this.parser = getOrDefault(parser, new AutoDetectParser());
+        this.contentHandler = getOrDefault(contentHandler, new BodyContentHandler(UNLIMITED));
+        this.metadata = getOrDefault(metadata, new Metadata());
+        this.parseContext = getOrDefault(parseContext, new ParseContext());
+    }
+
+    // TODO allow automatically extract metadata (e.g. creator, last-author, created/modified timestamp, etc)
+
+    @Override
+    public Document parse(InputStream inputStream) {
+        try {
+            parser.parse(inputStream, contentHandler, metadata, parseContext);
+            String text = contentHandler.toString();
```

Comment:
Have a look at https://github.com/quarkiverse/quarkus-tika/blob/main/runtime/src/main/java/io/quarkus/tika/TikaParser.java at how the embedded content from the images etc can be collected as well. It is not essential for this PR but can be useful in the future

Reply:
@sberyozkin sorry, my comment was in pending state 😬